File: solutions/java/say/1/src/main/java/Say.java
class Say {

    private static final String[] ONES = {
        "zero", "one", "two", "three", "four",
        "five", "six", "seven", "eight", "nine",
        "ten", "eleven", "twelve", "thirteen", "fourteen",
        "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"
    };

    private static final String[] TENS = {
        "", "", "twenty", "thirty", "forty",
        "fifty", "sixty", "seventy", "eighty", "ninety"
    };

    String say(long number) {

        if (number < 0 || number > 999999999999L) {
            throw new IllegalArgumentException();
        }

        if (number == 0) {
            return "zero";
        }

        return convert(number).trim();
    }

    private String convert(long number) {

        if (number < 20) {
            return ONES[(int) number];
        }

        if (number < 100) {
            return TENS[(int) (number / 10)]
                    + (number % 10 == 0 ? "" : "-" + convert(number % 10));
        }

        if (number < 1000) {
            return convert(number / 100) + " hundred"
                    + (number % 100 == 0 ? "" : " " + convert(number % 100));
        }

        if (number < 1_000_000) {
            return convert(number / 1000) + " thousand"
                    + (number % 1000 == 0 ? "" : " " + convert(number % 1000));
        }

        if (number < 1_000_000_000) {
            return convert(number / 1_000_000) + " million"
                    + (number % 1_000_000 == 0 ? "" : " " + convert(number % 1_000_000));
        }

        return convert(number / 1_000_000_000) + " billion"
                + (number % 1_000_000_000 == 0 ? "" : " " + convert(number % 1_000_000_000));
    }
}

File: solutions/python/say/1/say.py
ONES = {
    0: "zero", 1: "one", 2: "two", 3: "three", 4: "four",
    5: "five", 6: "six", 7: "seven", 8: "eight", 9: "nine",
    10: "ten", 11: "eleven", 12: "twelve", 13: "thirteen",
    14: "fourteen", 15: "fifteen", 16: "sixteen",
    17: "seventeen", 18: "eighteen", 19: "nineteen",
}

TENS = {
    20: "twenty", 30: "thirty", 40: "forty", 50: "fifty",
    60: "sixty", 70: "seventy", 80: "eighty", 90: "ninety",
}

SCALES = [
    (1_000_000_000, "billion"),
    (1_000_000, "million"),
    (1_000, "thousand"),
    (1, "")
]


def say_under_1000(n):
    """Convert a number < 1000 into words."""
    words = []

    # Hundreds
    if n >= 100:
        words.append(ONES[n // 100] + " hundred")
        n %= 100
        if n == 0:
            return " ".join(words)

    # 0–19
    if n < 20:
        if n != 0:
            words.append(ONES[n])
        return " ".join(words)

    # 20–99
    tens_part = (n // 10) * 10
    ones_part = n % 10

    if ones_part == 0:
        words.append(TENS[tens_part])
    else:
        words.append(TENS[tens_part] + "-" + ONES[ones_part])

    return " ".join(words)


def say(number):
    """Convert number to English words."""
    if number < 0 or number > 999_999_999_999:
        raise ValueError("input out of range")

    if number == 0:
        return "zero"

    result = []

    for scale_value, scale_name in SCALES:
        if number >= scale_value:
            chunk = number // scale_value
            number %= scale_value
            result.append(say_under_1000(chunk))
            if scale_name:
                result.append(scale_name)

    return " ".join(result)
